File: Python/ByTier/G/2114_G_Build_an_Array.py
# Problem: CF 2114 G - Build an Array
# https://codeforces.com/contest/2114/problem/G

"""
G. Build an Array

Problem Description:
Given a final array of length n and k operations, determine whether it's possible to construct the array by repeatedly adding elements and merging adjacent equal elements (which sum up). The key insight is that elements that are merged are powers of 2, and their contribution to the total number of operations can be tracked.

Approach:
- For each element in the array, decompose it into its odd part and power-of-2 factor.
- Use prefix and suffix arrays to track how many operations would be needed if we start/stop merging at certain indices.
- The maximum possible contribution of an element to the total count is calculated by considering the best prefix and suffix combinations.
- Time and Space Complexity:
    - Time: O(n) per test case, where n is the length of the array.
    - Space: O(n) for storing auxiliary arrays.

Algorithm:
1. Decompose each number into a power-of-2 factor (b) and an odd part (a).
2. Use two passes to compute prefix (l) and suffix (r) contributions of operations needed for merging.
3. For each element, calculate how it can contribute to total operations in the best way.
4. Compare with k to decide YES or NO.

"""

import sys

input = lambda: sys.stdin.readline().rstrip()
for _ in range(int(input())):
    n, k = map(int, input().split())
    w = list(map(int, input().split()))
    
    # a stores odd parts of numbers; b stores powers of 2
    a = []
    b = []
    for i in range(n):
        c, x = 0, w[i]
        while x % 2 == 0:
            x //= 2
            c += 1
        a.append(x)
        b.append(2**c)
    
    # l[i] stores prefix sum of operation counts from left
    l = [0] * (n + 1)
    # r[i] stores suffix sum of operation counts from right
    r = [0] * (n + 1)
    
    x, c = a[0], 0
    
    # Compute prefix contribution
    for i in range(n):
        l1 = b[i]
        if a[i] == x and i > 0 and b[i - 1] < b[i]:
            l1 -= 2 * b[i - 1] - 1  # Adjust when merging happens
        if a[i] != x:
            x = a[i]
        l[i] += l[i - 1] + l1
    
    l[n] = l[n - 1]  # Set l[n] to last valid value
    
    x = a[-1]
    
    # Compute suffix contribution
    for i in range(n - 1, -1, -1):
        r1 = b[i]
        if a[i] == x and i < n - 1 and b[i] > b[i + 1]:
            r1 -= 2 * b[i + 1] - 1  # Adjust when merging happens
        if a[i] != x:
            x = a[i]
        r[i] += r[i + 1] + r1
    
    r[-1] = r[0]  # Set r[-1] to first value for correct calculation
    
    # Final check: find maximum possible total contribution of any element
    for i in range(n):
        c = max(c, l[n] - l[i] + r[-1] - r[i] + b[i])
    
    print(["NO", "YES"][c >= k])


# https://github.com/VaHiX/codeForces/

File: Kotlin/ByRound/2114/2114_G_Build_an_Array.kt
// Problem: CF 2114 G - Build an Array
// https://codeforces.com/contest/2114/problem/G

/*
 * Problem: G. Build an Array
 * Purpose: Determine if a sequence of k operations can produce the given array from an empty array.
 *          Each operation adds an integer to either end of the array; adjacent identical elements merge into their sum.
 * 
 * Algorithm:
 * - For each element in the input array, calculate its "base" value (the odd part) and power of 2 (how many times it was halved).
 * - Use a sliding window technique to simulate possible merging sequences.
 * - Maintain bounds on how many operations are needed to construct the array based on cumulative sum computations.
 * 
 * Time Complexity: O(n) per test case, where n is the length of the array
 * Space Complexity: O(n), for storing intermediate arrays and computations
 */

private tailrec fun calc(d: Int, c: Int = 0): Pair<Int, Int> = if (d % 2 == 0) calc(d / 2, c + 1) else d to c

fun main() = java.util.Scanner(System.`in`).run {
    repeat(nextInt()) {
        val (n, k) = nextInt() to nextLong()
        List(n) { calc(nextInt()) }.unzip().let { (d, e) ->
            var p = d[0]
            // Initial computation of minimal operations for first element
            val init = (1 until n).fold(1L shl e[0]) { r, i -> 
                r + if (d[i] != p) { 
                    p = d[i]; 1L shl e[i] 
                } else if (e[i] > e[i - 1]) 
                    (1L shl e[i]) - (1L shl (e[i - 1] + 1)) + 1 
                else 
                    1L shl e[i] 
            }
            var res = init
            var q = d[0]
            // Compute maximum possible operations needed for valid sequence
            val final = (1 until n).fold(init) { a, i ->
                res = if (d[i] != q) { 
                    q = d[i]; 
                    (res - (1L shl e[i - 1]) + (1L shl e[i])) + (1L shl e[i - 1]) - (1L shl e[i]) 
                }
                else if (e[i] > e[i - 1]) 
                    (res - (1L shl e[i - 1]) + (1L shl e[i])) - ((1L shl e[i]) - (1L shl (e[i - 1] + 1)) + 1) + (1L shl e[i - 1])
                else 
                    (res - (1L shl e[i - 1]) + (1L shl e[i])) - (1L shl e[i]) + ((1L shl e[i - 1]) - (1L shl (e[i] + 1)) + 1)
                maxOf(a, res)
            }
            // Output YES if k is within valid range
            println(if (k in n..final) "YES" else "NO")
        }
    }
}


// https://github.com/VaHiX/CodeForces/